src/order/hom/lattice.lean

```diff
@@ -289,25 +365,209 @@ lemma cancel_left {g : inf_hom β γ} {f₁ f₂ : inf_hom α β} (hg : injectiv
+
+instance : inf_top_hom_class (inf_top_hom α β) α β :=
+{ coe :=  λ f, f.to_fun,
+  coe_injective' := λ f g h, by { obtain ⟨⟨_, _⟩, _⟩ := f, obtain ⟨⟨_, _⟩, _⟩ := g, congr' },
+  map_inf := λ f, f.map_inf',
+  map_top := λ f, f.map_top' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly. -/
+instance : has_coe_to_fun (inf_top_hom α β) (λ _, α → β) := fun_like.has_coe_to_fun
+
+@[simp] lemma to_fun_eq_coe {f : inf_top_hom α β} : f.to_fun = (f : α → β) := rfl
+
+@[ext] lemma ext {f g : inf_top_hom α β} (h : ∀ a, f a = g a) : f = g := fun_like.ext f g h
+
+/-- Copy of an `inf_top_hom` with a new `to_fun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy (f : inf_top_hom α β) (f' : α → β) (h : f' = f) : inf_top_hom α β :=
+{ to_inf_hom := f.to_inf_hom.copy f' h, ..f.to_top_hom.copy f' h }
+
+variables (α)
+
+/-- `id` as an `inf_top_hom`. -/
+@[simps] protected def id : inf_top_hom α α := ⟨inf_hom.id α, rfl⟩
+
+instance : inhabited (inf_top_hom α α) := ⟨inf_top_hom.id α⟩
+
+@[simp] lemma coe_id : ⇑(inf_top_hom.id α) = id := rfl
+
+variables {α}
+
+@[simp] lemma id_apply (a : α) : inf_top_hom.id α a = a := rfl
+
+/-- Composition of `inf_top_hom`s as an `inf_top_hom`. -/
+def comp (f : inf_top_hom β γ) (g : inf_top_hom α β) : inf_top_hom α γ :=
+{ ..f.to_inf_hom.comp g.to_inf_hom, ..f.to_top_hom.comp g.to_top_hom }
+
+@[simp] lemma coe_comp (f : inf_top_hom β γ) (g : inf_top_hom α β) : (f.comp g : α → γ) = f ∘ g :=
+rfl
+@[simp] lemma comp_apply (f : inf_top_hom β γ) (g : inf_top_hom α β) (a : α) :
+  (f.comp g) a = f (g a) := rfl
+@[simp] lemma comp_assoc (f : inf_top_hom γ δ) (g : inf_top_hom β γ) (h : inf_top_hom α β) :
+  (f.comp g).comp h = f.comp (g.comp h) := rfl
+@[simp] lemma comp_id (f : inf_top_hom α β) : f.comp (inf_top_hom.id α) = f := ext $ λ a, rfl
+@[simp] lemma id_comp (f : inf_top_hom α β) : (inf_top_hom.id β).comp f = f := ext $ λ a, rfl
+
+lemma cancel_right {g₁ g₂ : inf_top_hom β γ} {f : inf_top_hom α β} (hf : surjective f) :
+  g₁.comp f = g₂.comp f ↔ g₁ = g₂ :=
+⟨λ h, ext $ hf.forall.2 $ fun_like.ext_iff.1 h, congr_arg _⟩
+
+lemma cancel_left {g : inf_top_hom β γ} {f₁ f₂ : inf_top_hom α β} (hg : injective g) :
+  g.comp f₁ = g.comp f₂ ↔ f₁ = f₂ :=
+⟨λ h, inf_top_hom.ext $ λ a, hg $
+  by rw [←comp_apply, h, comp_apply], congr_arg _⟩
+
+end has_inf
+
+variables [semilattice_inf β] [order_top β]
+
+instance : has_inf (inf_top_hom α β) :=
+⟨λ f g, { to_inf_hom := f.to_inf_hom ⊓ g.to_inf_hom, ..f.to_top_hom ⊓ g.to_top_hom }⟩
+
+instance : semilattice_inf (inf_top_hom α β) :=
+fun_like.coe_injective.semilattice_inf _ $ λ f g, rfl
+
+instance : order_top (inf_top_hom α β) := { top := ⟨⊤, rfl⟩, le_top := λ f, le_top }
+
+@[simp] lemma coe_inf (f g : inf_top_hom α β) : ⇑(f ⊓ g) = f ⊓ g := rfl
+@[simp] lemma coe_top : ⇑(⊤ : inf_top_hom α β) = ⊤ := rfl
+@[simp] lemma inf_apply (f g : inf_top_hom α β) (a : α) : (f ⊓ g) a = f a ⊓ g a := rfl
```

Comment:
Can we have that these maps preserve `finset.inf` as well?

Reply:
What do you mean exactly? That `f (s.inf g) = s.inf (f ∘ g)` when `f : inf_top_hom α β` or that `⇑(s.sup f) = s.sup (λ i, ⇑(f i))` when `f : ι → inf_top_hom α β`?
Arguably, both lemmas go against the new `fun_like` design, as those results should respectively be accessed through
* a lemma about `inf_hom_class` that we're indeed missing
* the bundling of `coe_fn : inf_top_hom α β → α → β` as an `inf_hom` along with the aforementioned lemma.

Reply:
I mean some version of the former. I don't particularly mind how it's expressed but we should have some way to express this notion, especially because the PR name here talks about preserving finitary join, but none of your lemmas actually do this: they're all about the nullary or binary versions.

Reply:
I put them in `order.hom.lattice`, but arguably the import could go the other way. Haven't made up my mind.